### DISCUSSION TITLE
Improve display of instance references in the debugger

### DISCUSSION THREAD

**Discussion by mcerdeira** [godotengine/godot-proposals#5757] [Editor]
### Describe the project you are working on

Using Godot
Sample Project: [New Game Project.zip](https://github.com/godotengine/godot-proposals/files/9974453/New.Game.Project.zip)

### Describe the problem or limitation you are having in your project

Consider this scenario:

```
if enemy_obj: #check is the instance is alive (the wrong way!)
    enemy_obj.position.x += 1000 * delta
```

If a previously existing instaqnce (enemy_obj) is disposed using queue_free() this line fails because enemy_obj is NOT null, but also is no longer valid.
This is fine and works as expected, but, in the debugger, what you see is a little confusing:
<img width="561" alt="Captura" src="https://user-images.githubusercontent.com/938346/200920446-7421185f-f35b-41c8-bf3d-fb58ea671496.PNG">

The if statement is true, but in the inspector you see that enemy_obj tagged as null.

Real null error:
<img width="696" alt="Real null" src="https://user-images.githubusercontent.com/938346/200921668-ebdab4e4-d6e2-4edd-a926-baf967337779.PNG">

Invalid
<img width="690" alt="Invalid" src="https://user-images.githubusercontent.com/938346/200921906-3c12702b-ba72-4af8-8735-3b79ba6538b6.PNG">

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Make a distinction in the debbuger between null and invalid pointers.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

May be, show something like:

[Invalid]  or [freed instance] or something else thats != null.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

IDK

### Is there a reason why this should be core and not an add-on in the asset library?

IDK

**Comment by TheDuriel** [comment]
Null in this case stands for NullInstance, so it should read [NullInstance].

**Comment by mcerdeira** [comment]
I don't see why they closed the issue. 
In my opinion, is displaying data in a weird, confusing and counter intuitive way. Ambiguous at least.

**Comment by noidexe** [comment]
This is on Godot 4
![image](https://user-images.githubusercontent.com/526829/200970945-4b570ee5-be6e-4c54-a238-e75b13aacb14.png)

In the console it shows as `<Freed Object>`
![image](https://user-images.githubusercontent.com/526829/200971004-6509330d-d292-491d-b4d3-38d9acf72fe1.png)

But in the debugger it appears as `<null>`
![image](https://user-images.githubusercontent.com/526829/200970971-75bbf5f1-0b03-475e-a3d3-53f5ce07f876.png)

On Godot 3 it says `[Deleted Object]` instead of `Null`

It is indeed confusing. `<null>` is a concrete value a variable can have, and it's falsy while `<Freed Object>` is truthy. 
If print() can tell the difference why can't the debugger do the same? I assume this is a limitation rather than a bug, right? I can't find where the code for showing stack variables is.

**Comment by TheDuriel** [comment]
Sounds like you'll want to make a bug report issue then! That's a regression for sure.

**Comment by noidexe** [comment]
It's already reported as a bug: https://github.com/godotengine/godot/issues/49019

**Comment by TheDuriel** [comment]
No point in this "discussion" then. Gotta wait for a fix.

**Reply by noidexe** [reply]
I don't get what you mean with the "". This was moved to discussions from proposals. To make it clear, I've just found it's reported as a bug. I don't think it's a regression though. I've never seen the debugger display "deleted/freed object" so it makes sense to assume it was like this by design or by some limitation. 
But well, if it was already accepted as a bug then that's even better than a proposal.
